zenith_utils/src/lsn.rs

```diff
@@ -0,0 +1,175 @@
+#![warn(missing_docs)]
+
+use std::fmt;
+use std::ops::{Add, AddAssign};
+use std::path::Path;
+use std::str::FromStr;
+use std::sync::atomic::{AtomicU64, Ordering};
+
+// FIXME: should we implement conversions to/from PgLsn?
+// Doing so would mean zenith_utils gets a dependency on postgres-types.
+// That's not necessarily a bad thing, but if there is only one place
+// that needs it, it might be easier to do it there, e.g.
+//     let x = Lsn::from(u64::from(pglsn))
+
+/// A Postgres LSN (Log Sequence Number), also known as an XLogRecPtr
+#[derive(Debug, Clone, Copy, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Lsn(pub u64);
+
+/// We tried to parse an LSN from a string, but failed
+#[derive(Debug, PartialEq, thiserror::Error)]
+#[error("LsnParseError")]
+pub struct LsnParseError;
+
+impl Lsn {
+    /// Maximum possible value for an LSN
+    pub const MAX: Lsn = Lsn(u64::MAX);
+
+    /// Subtract a number, returning None on overflow.
+    pub fn checked_sub<T: Into<u64>>(self, other: T) -> Option<Lsn> {
+        let other: u64 = other.into();
+        self.0.checked_sub(other).map(Lsn)
+    }
+
+    /// Parse an LSN from a filename in the form `0000000000000000`
+    pub fn from_filename<F>(filename: F) -> Result<Self, LsnParseError>
+    where
+        F: AsRef<Path>,
+    {
+        let filename: &Path = filename.as_ref();
+        let filename = filename.to_str().ok_or(LsnParseError)?;
+        let n = u64::from_str_radix(filename, 16).or(Err(LsnParseError))?;
+        Ok(Lsn(n))
+    }
+
+    // FIXME: Is there a more ergonomic way to do this?
+    //
+    // For example, we could have a SegmentGeometry struct that
+    // implements this logic instead, e.g.
+    //
+    //    const seg_geo = SegmentGeometry::new(16 * 1024 * 1024);
+    //    let offset = seg_geo.seg_offset(lsn);
+    //    let seg_no = seg_geo.seg_num(lsn);
+
+    /// Compute the offset into a segment
+    pub fn segment_offset(self, seg_sz: u64) -> u64 {
+        self.0 % seg_sz
+    }
```

Comment:
Does anyone have an opinion about this?

I don't see many places where segment size needs to be acted on, so I'll probably leave this as-is unless someone wants me to be more ambitious...

Reply:
IMHO "segment geometry" seems to be confusing. Actually segment size is specified during Postgres configuration and page server/wal acceptor receives this infirmation from from computation node.